stdlib/asm/mem.masm

```diff
@@ -49,3 +49,132 @@ export.memcopy
   # stack: [...]
   dropw drop drop drop
 end
+
+#! Copies an even number of words from the advice_stack to memory.
+#!
+#! Input: [C, B, A, write_ptr, num_words, ...]
+#! Output: [C, B, A, write_ptr, ...]
+#!
+#! Where:
+#! - The words C, B, and A are the RPO hasher state
+#!     - A is the capacity
+#!     - C,B are the rate portion of the state
+#! - num_words must be even and greater than zero
+#!
+#! Cycles: 8 + 11 * num_words
+export.move_even_words_from_advice_stack_to_memory.0
+  dup.13 neq.0 # (3 cycles)
+
+  # loop until counter is zero (11 cycles per iteration)
+  # LOOP: [b, C, B, A, write_ptr, num_words, ...]
+  while.true
+    adv_pipe # (2 cycles)
+    # => [C', B', A', write_ptr', num_words, ...]
+
+    # update `num_words`, it is required to be even
+    swapw.3 swap sub.2 swap swapw.3 # (6 cycles)
+    # => [C', B', A', write_ptr', num_words-2, ...]
+
+    dup.13 neq.0 # (3 cycles)
+    # LOOP: [C', B', A', write_ptr', num_words-2, ...]
+  end
+
+  movup.13 drop # (5 cycles)
+  # [C', B', A', write_ptr', ...]
+end
+
+#! Copies an arbitrary number of words from the advice tape to memory
+#!
+#! Input: [num_words, write_ptr, ...]
+#! Output: [HASH, write_ptr', ...]
+#! Cycles:
+#!  even num_words: 41 + 8 * num_words
+#!  odd num_words: 58 + 8 * num_words
+export.move_words_from_advice_stack_to_memory.0
```

Comment:
Maybe we should name this `pipe_words_to_memory`?

Reply:
renamed